### PR TITLE
Adding support for wandb

### PR DESCRIPTION
**Pull request by manangoel99**
This PR fixes issues #1030 and #1049 . It adds support for the wandb logger.
I've added a new command line argument `--logger` which accepts "tensorboard" and "wandb" as arguments and raises an error otherwise.
More wandb related arguments like `project`, `entity` etc. can be provided in the CLI using the prefix `wandb-`. For example

```shell
python tools/train.py -n yolox-s -d 8 -b 64 --fp16 -o [--cache] --logger wandb wandb-project <project name>
                         yolox-m
                         yolox-l
                         yolox-x
```

All the metrics are logged to the W&B dashboard along with the hyperparameters. The model checkpoints are stored as wandb artifacts along with tagging the best and latest models appropriately.

[This dashboard](https://wandb.ai/manan-goel/yolox-s/runs/1mg3zu90) uses yolox-tiny.

### CODE REVIEW

**Comment by FateScript** [comment]
Thanks for your contribution @manangoel99 Plz lint your code to pass the workflow.

**Comment by manangoel99** [comment]
> Thanks for your contribution @manangoel99 Plz lint your code to pass the workflow.

Fixed thanks!

File: docs/quick_run.md [hunk 1]
```diff
@@ -56,6 +56,26 @@ python tools/train.py -n yolox-s -d 8 -b 64 --fp16 -o [--cache]
...
+pip install wandb
+wandb login
+```
+
+Log in to your W&B account
+
+To start logging metrics to W&B during training add the flag `--logger` to the previous command and use the prefix "wandb-" to specify arguments for initializing the wandb run.
+
+```shell
+python tools/train.py -n yolox-s -d 8 -b 64 --fp16 -o [--cache] --logger wandb wandb-project <project name>
```
Comment (FateScript) [inline docs/quick_run.md:73]: What if using cli like `python tools/train.py -n yolox-s -d 8 -b 64 --fp16 -o --logger wandb wandb-project <project name> data_num_workers 0`?

File: yolox/core/trainer.py [hunk 2]
```diff
@@ -334,3 +360,6 @@ def save_ckpt(self, ckpt_name, update_best_ckpt=False):
                 self.file_name,
                 ckpt_name,
             )
+
+            if self.args.logger == 'wandb':
```
Comment (FateScript) [inline yolox/core/trainer.py:364]: please use `'` instead of `"` for normal string.

File: yolox/utils/logger.py [hunk 3]
```diff
@@ -93,3 +95,122 @@ def setup_logger(save_dir, distributed_rank=0, filename="log.txt", mode="a"):
...
+        _ = self.run
+
+        if self.config:
+            self.run.config.update(self.config)
+        self.run.define_metric("epoch")
+        self.run.define_metric("val/", step_metric="epoch")
+
+    @property
+    def run(self):
+        if self._run is None:
```
Comment (FateScript) [inline yolox/utils/logger.py:165]: why not using `self.run` as attribute but as property?

**Review by FateScript — COMMENTED** [review]
PTAL @manangoel99 , please also use isort=4.3.21 to sort import of your code, thanks a lot.

File: docs/quick_run.md (hunk 1 above)
Comment (manangoel99) [reply docs/quick_run.md:73]: This works. I iterate over `args.opts` to see if the prefix is `wandb-` to initialize the wandb logger.

File: yolox/core/trainer.py (hunk 2 above)
Comment (manangoel99) [reply yolox/core/trainer.py:364]: Fixed

File: yolox/utils/logger.py (hunk 3 above)
Comment (manangoel99) [reply yolox/utils/logger.py:165]: We use `self.run` as a property to make sure that the run object is only readable.

**Review by FateScript — APPROVED** [review]
LGTM

**Comment by samithaFHSS** [comment]
I'm getting an error when using wandb with custom data. Error is `'VOCDetection' object has no attribute 'cats'`. Any solution for this?

**Comment by scottire** [comment]
hi @samithaFHSS, have you been able to resolve the above issue? If not, can you share code to reproduce this issue please?

**Comment by Hyper-Devil** [comment]
@scottire Sorry to bother you, this is my log.

2022-08-12 21:29:26.829 | INFO     | yolox.core.trainer:before_train:136 - Model Summary: Params: 8.94M, Gflops: 26.76
2022-08-12 21:29:28.867 | INFO     | yolox.core.trainer:before_train:155 - init prefetcher, this might take one minute or less...
2022-08-12 21:29:50.369 | ERROR    | yolox.core.launch:_distributed_worker:147 - An error has been caught in function '_distributed_worker', process 'SpawnProcess-1' (78), thread 'MainThread' (140542829814976):
Traceback (most recent call last):

  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.9/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               │     │   └ 5
               │     └ 8
               └ <function _main at 0x7fd2ad313940>
  File "/usr/local/lib/python3.9/multiprocessing/spawn.py", line 129, in _main
    return self._bootstrap(parent_sentinel)
           │    │          └ 5
           │    └ <function BaseProcess._bootstrap at 0x7fd2ad41ca60>
           └ <SpawnProcess name='SpawnProcess-1' parent=11 started>
  File "/usr/local/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
    │    └ <function BaseProcess.run at 0x7fd2ad41c0d0>
    └ <SpawnProcess name='SpawnProcess-1' parent=11 started>
  File "/usr/local/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
    │    │        │    │        │    └ {}
    │    │        │    │        └ <SpawnProcess name='SpawnProcess-1' parent=11 started>
    │    │        │    └ (<function _distributed_worker at 0x7fd120eccee0>, 0, (<function main at 0x7fd11498f790>, 2, 2, 0, 'nccl', 'tcp://127.0.0.1:4...
    │    │        └ <SpawnProcess name='SpawnProcess-1' parent=11 started>
    │    └ <function _wrap at 0x7fd1212ad700>
    └ <SpawnProcess name='SpawnProcess-1' parent=11 started>
  File "/usr/local/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
    │  │   └ (<function main at 0x7fd11498f790>, 2, 2, 0, 'nccl', 'tcp://127.0.0.1:40017', (╒═══════════════════╤═════════════════════════...
    │  └ 0
    └ <function _distributed_worker at 0x7fd120eccee0>

> File "/root/YOLOX-docker/yolox/core/launch.py", line 147, in _distributed_worker
    main_func(*args)
    │          └ (╒═══════════════════╤═══════════════════════════════════════════════════════════════════════════════════════════════════════...
    └ <function main at 0x7fd11498f790>

  File "/root/YOLOX-docker/tools/train.py", line 118, in main
    trainer.train()
    │       └ <function Trainer.train at 0x7fd1143de4c0>
    └ <yolox.core.trainer.Trainer object at 0x7fd1143f4b80>

  File "/root/YOLOX-docker/yolox/core/trainer.py", line 74, in train
    self.before_train()
    │    └ <function Trainer.before_train at 0x7fd1143ded30>
    └ <yolox.core.trainer.Trainer object at 0x7fd1143f4b80>

  File "/root/YOLOX-docker/yolox/core/trainer.py", line 183, in before_train
    self.wandb_logger = WandbLogger.initialize_wandb_logger(
    │                   │           └ <classmethod object at 0x7fd11f38ae50>
    │                   └ <class 'yolox.utils.logger.WandbLogger'>
    └ <yolox.core.trainer.Trainer object at 0x7fd1143f4b80>

  File "/root/YOLOX-docker/yolox/utils/logger.py", line 384, in initialize_wandb_logger
    return cls(config=vars(exp), val_dataset=val_dataset, **wandb_params)
           │               │                 │              └ {'project': 'yolox-s'}
           │               │                 └ <yolox.data.datasets.voc.VOCDetection object at 0x7fd146cec5e0>
           │               └ ╒═══════════════════╤════════════════════════════════════════════════════════════════════════════════════════════════════════...
           └ <class 'yolox.utils.logger.WandbLogger'>

  File "/root/YOLOX-docker/yolox/utils/logger.py", line 206, in __init__
    self.cats = val_dataset.cats
    │           └ <yolox.data.datasets.voc.VOCDetection object at 0x7fd146cec5e0>
    └ <yolox.utils.logger.WandbLogger object at 0x7fd146f0ddc0>

AttributeError: 'VOCDetection' object has no attribute 'cats'

**Comment by manangoel99** [comment]
Hi @samithaFHSS ! The reason this error is coming up is because the logger only currently supports COCO datasets and not VOC datasets. We will try to add it ASAP

**Pull request merged**